File: weatherman/weatherman_application.py
import csv
from datetime import datetime


class WeathermanApplication:

    MONTHS = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]

    def __init__(self, path):
        self.path = path
        self.weather_data_list = []
        
    def read_year_files(self, year):
        for i in range(12):
            try:
                filename = 'Murree_weather_{}_{}.txt'.format(year, self.MONTHS[i])
                with open(self.path + filename) as csvfile:
                    temp_list = []
                    reader = csv.DictReader(csvfile)
                    for day_weather in reader:
                        temp_list.append(day_weather)
                    self.weather_data_list.append(temp_list)
            except Exception as e :
                print("This Year doesnt exist or " + str(i) + " Month is not available")
                print(e)
        
    def read_month_files(self, year,month):
        try:
            filename = 'Murree_weather_{}_{}.txt'.format(year, self.MONTHS[month-1])
            with open(self.path + filename) as csvfile:
                temp_list = []
                reader = csv.DictReader(csvfile)
                for day_weather in reader:
                    temp_list.append(day_weather)
                self.weather_data_list.append(temp_list)
        except Exception as e :
            print("This Year doesnt exist or " + str(month) + " Month is not available")
            print(e)

    def find_highest_lowest_temperature_and_max_humidity(self):
        if len(self.weather_data_list) == 0:
            print("No data is available")
        max_tempC = max(int(day_weather["Max TemperatureC"]) if day_weather["Max TemperatureC"] != "" else 0 for day_weather in self.weather_data_list[0])
        max_tempC_date = datetime.now()
        for month in self.weather_data_list:
            for day_weather in month:
                temp = int(day_weather["Max TemperatureC"]) if day_weather["Max TemperatureC"] != "" else None
                if temp is not None and temp > max_tempC:
                    max_tempC_date = datetime.strptime(day_weather["PKT"], "%Y-%m-%d").date()
                    max_tempC = temp
        print("Maximum Temperature:",max_tempC,"C" , max_tempC_date.strftime("%b %d"))

        min_tempC = min(int(day_weather["Min TemperatureC"]) if day_weather["Min TemperatureC"] != "" else 0 for day_weather in self.weather_data_list[0])
        min_tempC_date = datetime.now()
        for month in self.weather_data_list:
            for day_weather in month:
                temp = int(day_weather["Min TemperatureC"]) if day_weather["Min TemperatureC"] != "" else None
                if temp is not None and temp < min_tempC:
                    min_tempC_date = datetime.strptime(day_weather["PKT"], "%Y-%m-%d").date()
                    min_tempC = temp
        print("Minimum Temperature:",min_tempC,"C" , min_tempC_date.strftime("%b %d"))

        max_humidity = max(int(day_weather["Max Humidity"]) if day_weather["Max Humidity"] != "" else 0 for day_weather in self.weather_data_list[0])
        max_humidity_date = datetime.now()
        for month in self.weather_data_list:
            for day_weather in month:
                temp = int(day_weather["Max Humidity"]) if day_weather["Max Humidity"] != "" else None
                if temp is not None and temp > max_humidity:
                    max_humidity_date = datetime.strptime(day_weather["PKT"], "%Y-%m-%d").date()
                    max_humidity = temp
        print("Maximum Humidity:",max_humidity,"%" , max_humidity_date.strftime("%b %d"))
        print()
    
    def find_average_max_temp_low_temp_and_mean_humidity(self):
        if len(self.weather_data_list) == 0:
            print("No data is available")
        day_counter = 0
        max_tempC_sum = 0
        for day_weather in self.weather_data_list[0]:
            if day_weather["Max TemperatureC"] != "":
                max_tempC_sum += int(day_weather["Max TemperatureC"])
                day_counter += 1
        avg_max_tempC = round(max_tempC_sum/day_counter,2)
        print("Average Highest:",avg_max_tempC,"C")

        day_counter = 0
        min_tempC_sum = 0
        for day_weather in self.weather_data_list[0]:
            if day_weather["Min TemperatureC"] != "":
                min_tempC_sum += int(day_weather["Min TemperatureC"])
                day_counter += 1
        avg_min_tempC = round(min_tempC_sum/day_counter,2)
        print("Average Lowest:",avg_min_tempC,"C")

        day_counter = 0
        mean_humidity_sum = 0
        for day_weather in self.weather_data_list[0]:
            if day_weather[" Mean Humidity"] != "":
                mean_humidity_sum += int(day_weather[" Mean Humidity"])
                day_counter += 1
        avg_mean_humidity = round(mean_humidity_sum/day_counter,2)
        print("Average Mean Humdidity:",avg_mean_humidity,"%")
        print()

    def bar_chart_vertically(self):
        if len(self.weather_data_list) == 0:
            print("No data is available")
        counter = 1
        for day_weather in self.weather_data_list[0]:
            max_temp = int(day_weather["Max TemperatureC"]) if day_weather["Max TemperatureC"] != "" else 0
            min_temp = int(day_weather["Min TemperatureC"]) if day_weather["Min TemperatureC"] != "" else 0
            print('\x1b[3;31m' + str(counter), abs(max_temp) * "+", max_temp, "C \x1b[0m")
            print('\x1b[3;34m' + str(counter), abs(min_temp) * "+", min_temp, "C \x1b[0m")
            counter += 1

    def bar_chart_horizentally(self):
        if len(self.weather_data_list) == 0:
            print("No data is available")
        counter = 1
        for day_weather in self.weather_data_list[0]:
            max_temp = int(day_weather["Max TemperatureC"]) if day_weather["Max TemperatureC"] != "" else 0
            min_temp = int(day_weather["Min TemperatureC"]) if day_weather["Min TemperatureC"] != "" else 0
            print('\x1b[3;34m' + str(counter), abs(min_temp) * "+" + '\x1b[3;31m', abs(max_temp) * "+", str(min_temp) + " - " + str(max_temp)+"C \x1b[0m")
            counter += 1

File: WeathermanApplication.go
package main

import (
	"encoding/csv"
	"fmt"
	"io"
	"log"
	"os"
	"strconv"
	"time"
)

var MONTHS = [...]string{"jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"}

type weather_data_map_list [][]map[string]string

type WeathermanApplication struct {
	path              string
	weather_data_list weather_data_map_list
}

func (context *WeathermanApplication) Init(fpath string) {
	context.path = fpath
}

func (context *WeathermanApplication) ReadYearWeather(year string) {
	context.weather_data_list = [][]map[string]string{}
	for i := 0; i < 12; i++ {
		filepath := context.path + "/Murree_weather_" + year + "_" + MONTHS[i] + ".txt"
		file, err := os.Open(filepath)
		if err == nil {
			file_reader := csv.NewReader(file)
			file_reader.Comma = ','
			var header []string
			var new_month = []map[string]string{}
			for {
				record, err := file_reader.Read()
				if err == io.EOF {
					break
				}
				if err != nil {
					log.Fatal(err)
				}
				if header == nil {
					header = record
				} else {
					weather_data := map[string]string{}
					for i := range header {
						weather_data[header[i]] = record[i]
					}
					new_month = append(new_month, weather_data)
				}
			}
			context.weather_data_list = append(context.weather_data_list, new_month)
		}
	}
}

func (context *WeathermanApplication) ReadMonthWeather(year, month string) {
	context.weather_data_list = [][]map[string]string{}
	mon, _ := strconv.Atoi(month)
	filepath := context.path + "/Murree_weather_" + year + "_" + MONTHS[mon-1] + ".txt"
	file, err := os.Open(filepath)
	if err == nil {
		file_reader := csv.NewReader(file)
		file_reader.Comma = ','
		var header []string
		var new_month = []map[string]string{}
		for {
			record, err := file_reader.Read()
			if err == io.EOF {
				break
			}
			if err != nil {
				log.Fatal(err)
			}
			if header == nil {
				header = record
			} else {
				weather_data := map[string]string{}
				for i := range header {
					weather_data[header[i]] = record[i]
				}
				new_month = append(new_month, weather_data)
			}
		}
		context.weather_data_list = append(context.weather_data_list, new_month)
	}
}

func (context *WeathermanApplication) find_highest_lowest_temperature_and_max_humidity() {
	if len(context.weather_data_list) == 0 {
		fmt.Println("No Data is Available")
	} else {
		max_tempC := -999 //non real max value
		max_tempC_date := time.Now()
		min_tempC := 999
		min_tempC_date := time.Now()
		max_humidity := -999
		max_humidity_date := time.Now()
		for _, months := range context.weather_data_list {
			for _, day_weather := range months {
				layout := "2006-1-2"
				temp, err := strconv.Atoi(day_weather["Max TemperatureC"])
				if err == nil {
					if temp > max_tempC {
						max_tempC = temp
						max_tempC_date, _ = time.Parse(layout, day_weather["PKT"])
					}
				}
				temp, err = strconv.Atoi(day_weather["Min TemperatureC"])
				if err == nil {
					if temp < min_tempC {
						min_tempC = temp
						min_tempC_date, _ = time.Parse(layout, day_weather["PKT"])
					}
				}
				temp, err = strconv.Atoi(day_weather["Max Humidity"])
				if err == nil {
					if temp > max_humidity {
						max_humidity = temp
						max_humidity_date, _ = time.Parse(layout, day_weather["PKT"])
					}
				}
			}
		}
		fmt.Printf("Highest : %vC on %v %v", max_tempC, max_tempC_date.Month(), max_tempC_date.Day())
		fmt.Println()
		fmt.Printf("Lowest : %vC on %v %v", min_tempC, min_tempC_date.Month(), min_tempC_date.Day())
		fmt.Println()
		fmt.Printf("humidity : %vC on %v %v", max_humidity, max_humidity_date.Month(), max_humidity_date.Day())
		fmt.Println()
	}
}

func (context *WeathermanApplication) find_average_max_temp_low_temp_and_mean_humidity() {
	if len(context.weather_data_list) == 0 {
		fmt.Println("No Data is Available")
	} else {
		max_tempC_day_counter := 0
		min_tempC_day_counter := 0
		mean_humidity_day_counter := 0
		max_tempC_sum := 0
		min_tempC_sum := 0
		mean_humidity_sum := 0
		for _, day_weather := range context.weather_data_list[0] {
			temp, err := strconv.Atoi(day_weather["Max TemperatureC"])
			if err == nil {
				max_tempC_day_counter++
				max_tempC_sum += temp
			}
			temp, err = strconv.Atoi(day_weather["Min TemperatureC"])
			if err == nil {
				min_tempC_day_counter++
				min_tempC_sum += temp
			}
			temp, err = strconv.Atoi(day_weather[" Mean Humidity"])
			if err == nil {
				mean_humidity_day_counter++
				mean_humidity_sum += temp
			}
		}
		fmt.Printf("Highest Average : %vC", max_tempC_sum/max_tempC_day_counter)
		fmt.Println()
		fmt.Printf("Lowest Average : %vC", min_tempC_sum/min_tempC_day_counter)
		fmt.Println()
		fmt.Printf("Mean Humidity Average : %vC", mean_humidity_sum/mean_humidity_day_counter)
		fmt.Println()
	}
}

func (context *WeathermanApplication) bar_chart_vertically() {
	if len(context.weather_data_list) == 0 {
		fmt.Println("No Data is Available")
	} else {
		counter := 1
		for _, day_weather := range context.weather_data_list[0] {
			max_temp, err1 := strconv.Atoi(day_weather["Max TemperatureC"])
			min_tempC, err2 := strconv.Atoi(day_weather["Min TemperatureC"])
			if max_temp < 0 {
				max_temp *= -1
			}
			if min_tempC < 0 {
				min_tempC *= -1
			}
			if err1 == nil {
				plus := ""
				for i := 0; i < max_temp; i++ {
					plus += "+"
				}
				fmt.Println("\x1b[3;31m ", counter, plus, max_temp, "C \x1b[0m")
			}
			if err2 == nil {
				plus := "+"
				for i := 0; i < min_tempC; i++ {
					plus += "+"
				}
				fmt.Println("\x1b[3;34m ", counter, plus, min_tempC, "C \x1b[0m")
			}
			counter++
		}
	}
}

func (context *WeathermanApplication) bar_chart_horizentally() {
	if len(context.weather_data_list) == 0 {
		fmt.Println("No Data is Available")
	} else {
		counter := 1
		for _, day_weather := range context.weather_data_list[0] {
			max_tempC, err1 := strconv.Atoi(day_weather["Max TemperatureC"])
			min_tempC, err2 := strconv.Atoi(day_weather["Min TemperatureC"])
			if max_tempC < 0 {
				max_tempC *= -1
			}
			if min_tempC < 0 {
				min_tempC *= -1
			}
			if err1 == nil && err2 == nil {
				max_plus := ""
				for i := 0; i < max_tempC; i++ {
					max_plus += "+"
				}
				min_plus := ""
				for i := 0; i < min_tempC; i++ {
					min_plus += "+"
				}
				fmt.Println("\x1b[3;34m ", counter, min_plus, "\x1b[3;31m ", max_plus, min_tempC, "-", max_tempC, "C \x1b[0m")
			}
			counter++
		}
	}
}
